building 1.1.0 from source failed on OSX

Hi,

Looking forward to using nixd! I tried adding it to my home-manager config and `switch`ing failed because compiling nixd 1.1.0 failed. I had to allow broken package for the system to try building stuff, so I guess that's normal?

My nix-info:
```md
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.13.3`
 - channels(auguste): `"darwin, home-manager"`
 - nixpkgs: `/nix/store/q336zlh9g3f5k0vabn6xzwhxasnhk7hv-source`
```

Logs:
```
building '/nix/store/r8jcz20nq15iz3nr68hnncbxk7mldkhn-nixd-1.1.0.drv'...
unpacking sources
unpacking source archive /nix/store/xn024694fg0ib0qxd7z7hw4rqas2n1rs-source
source root is source
patching sources
configuring
meson flags: --buildtype=release         --libdir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/lib --libexecdir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/libexec         --bindir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/bin --sbindir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/sbin         --includedir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/include         --mandir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/man --infodir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/info         --localedir=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/locale         -Dauto_features=enabled         -Dwrap_mode=nodownload         --prefix=/nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0
The Meson build system
Version: 1.1.0
Source dir: /private/tmp/nix-build-nixd-1.1.0.drv-0/source
Build dir: /private/tmp/nix-build-nixd-1.1.0.drv-0/source/build
Build type: native build
Project name: nixd
Project version: 1.1.0
C compiler for the host machine: clang (clang 11.1.0 "clang version 11.1.0")
C linker for the host machine: clang ld64 609
C++ compiler for the host machine: clang++ (clang 11.1.0 "clang version 11.1.0")
C++ linker for the host machine: clang++ ld64 609
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program git found: NO
Configuring nixd-config.h using configuration
Found pkg-config: /nix/store/l7np09c9h53qgk7jr2g4fwa72pmsrhlj-pkg-config-wrapper-0.29.2/bin/pkg-config (0.29.2)
Run-time dependency GTest found: YES 1.12.1
Run-time dependency gtest_main found: YES 1.12.1
Did not find CMake 'cmake'
Found CMake: NO
llvm-config found: YES (/nix/store/0a6r07k90smz2rmrscfnkg4hi7nfj02z-llvm-16.0.1-dev/bin/llvm-config) 16.0.1
Run-time dependency LLVM found: YES 16.0.1
Run-time dependency Boost found: YES 1.82.0 (/nix/store/qvy51344bxgg2sys3s4861ax1lq8k40n-boost-1.82.0-dev)
Library backtrace found: YES
Library nixfetchers found: YES
Run-time dependency nix-expr found: YES 2.16.1
Run-time dependency nix-cmd found: YES 2.16.1
Run-time dependency nix-store found: YES 2.16.1
Run-time dependency nix-main found: YES 2.16.1
Program flex found: YES (/nix/store/15c0wzfbsr7cy1n6vjc63f8jz1rr7pfk-flex-2.6.4/bin/flex)
Program bison found: YES (/nix/store/87znz73b3afw64yj2sqc95w61n1glpr8-bison-3.8.2/bin/bison)
Program lit found: YES (/nix/store/c050yai2vhfmwzx57igjyf88q2w0yj8w-python3.10-lit-15.0.6/bin/lit)
Build targets in project: 7

nixd 1.1.0

  User defined options
    auto_features: enabled
    bindir       : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/bin
    buildtype    : release
    includedir   : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/include
    infodir      : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/info
    libdir       : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/lib
    libexecdir   : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/libexec
    localedir    : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/locale
    mandir       : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/share/man
    prefix       : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0
    sbindir      : /nix/store/nghgypvc24lqpsibldfj4gh9ypxfsgnp-nixd-1.1.0/sbin
    wrap_mode    : nodownload

Found ninja-1.11.1 at /nix/store/npzka7a63hkzj9z8cq0a8vms4ldx8cq3-ninja-1.11.1/bin/ninja
meson: enabled parallel building
meson: enabled parallel installing
building
build flags: -j4
[1/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o
FAILED: lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o
clang++ -Ilib/lspserver/libnixd-lspserver.dylib.p -Ilib/lspserver -I../lib/lspserver -I../lib/lspserver/include -I/nix/store/0a6r07k90smz2rmrscfnkg4hi7nfj02z-llvm-16.0.1-dev/include -I/private/tmp/nix-build-nixd-1.1.0.drv-0/source/build -fcolor-diagnostics -Wall -Winvalid-pch -std=gnu++20 -O3 -include /nix/store/5l17m6jv3c1qx0drzga4ik6vxr0aazw2-nix-2.16.1-dev/include/nix/config.h -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MQ lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o -MF lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o.d -o lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o -c ../lib/lspserver/src/Logger.cpp
../lib/lspserver/src/Logger.cpp:104:38: error: use of undeclared identifier 'PTHREAD_MUTEX_ROBUST'
  pthread_mutexattr_setrobust(&Attr, PTHREAD_MUTEX_ROBUST);
                                     ^
1 error generated.
[2/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_DraftStore.cpp.o
[3/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_LSPServer.cpp.o
[4/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_Connection.cpp.o
ninja: build stopped: subcommand failed.
error: builder for '/nix/store/r8jcz20nq15iz3nr68hnncbxk7mldkhn-nixd-1.1.0.drv' failed with exit code 1;
       last 10 log lines:
       > FAILED: lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o
       > clang++ -Ilib/lspserver/libnixd-lspserver.dylib.p -Ilib/lspserver -I../lib/lspserver -I../lib/lspserver/include -I/nix/store/0a6r07k90smz2rmrscfnkg4hi7nfj02z-llvm-16.0.1-dev/include -I/private/tmp/nix-build-nixd-1.1.0.drv-0/source/build -fcolor-diagnostics -Wall -Winvalid-pch -std=gnu++20 -O3 -include /nix/store/5l17m6jv3c1qx0drzga4ik6vxr0aazw2-nix-2.16.1-dev/include/nix/config.h -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MQ lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o -MF lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o.d -o lib/lspserver/libnixd-lspserver.dylib.p/src_Logger.cpp.o -c ../lib/lspserver/src/Logger.cpp
       > ../lib/lspserver/src/Logger.cpp:104:38: error: use of undeclared identifier 'PTHREAD_MUTEX_ROBUST'
       >   pthread_mutexattr_setrobust(&Attr, PTHREAD_MUTEX_ROBUST);
       >                                      ^
       > 1 error generated.
       > [2/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_DraftStore.cpp.o
       > [3/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_LSPServer.cpp.o
       > [4/45] Compiling C++ object lib/lspserver/libnixd-lspserver.dylib.p/src_Connection.cpp.o
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/r8jcz20nq15iz3nr68hnncbxk7mldkhn-nixd-1.1.0.drv'.
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -46,9 +46,9 @@ jobs:
     - uses: cachix/install-nix-action@v20
       with:
         github_access_token: ${{ secrets.GITHUB_TOKEN }}
-    - run: nix develop --command env CXX=clang++ meson setup build/ --buildtype=${{ matrix.buildtype }} -Db_sanitize=${{ matrix.sanitizer }} -Db_ndebug=${{ matrix.ndebug }}
-    - run: nix develop --command meson compile -C build
-    - run: nix develop --command meson test -C build
+    - run: nix develop .#llvm --command env CXX=clang++ meson setup build/ --buildtype=${{ matrix.buildtype }} -Db_sanitize=${{ matrix.sanitizer }} -Db_ndebug=${{ matrix.ndebug }}
+    - run: nix develop .#llvm --command meson compile -C build
+    - run: nix develop .#llvm --command meson test -C build
     - uses: actions/upload-artifact@v3
       if: failure()
       with:

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -18,7 +18,7 @@ jobs:
     if: ${{ !contains(github.event.head_commit.message, '[skip ci]') }}
     steps:
     - uses: actions/checkout@v3
-    - uses: cachix/install-nix-action@v20
+    - uses: cachix/install-nix-action@v22
       with:
         github_access_token: ${{ secrets.GITHUB_TOKEN }}
     - run: nix develop --command meson setup build/ --buildtype=${{ matrix.buildtype }} -Db_sanitize=${{ matrix.sanitizer }} -Db_ndebug=${{ matrix.ndebug }}
@@ -43,7 +43,7 @@ jobs:
     if: ${{ !contains(github.event.head_commit.message, '[skip ci]') }}
     steps:
     - uses: actions/checkout@v3
-    - uses: cachix/install-nix-action@v20
+    - uses: cachix/install-nix-action@v22
       with:
         github_access_token: ${{ secrets.GITHUB_TOKEN }}
     - run: nix develop .#llvm --command env CXX=clang++ meson setup build/ --buildtype=${{ matrix.buildtype }} -Db_sanitize=${{ matrix.sanitizer }} -Db_ndebug=${{ matrix.ndebug }}

--- a/.github/workflows/nix.yml
+++ b/.github/workflows/nix.yml
@@ -7,7 +7,7 @@ jobs:
   flake-default:
     strategy:
       matrix:
-        os: [ubuntu-latest]
+        os: [ ubuntu-latest, macos-latest ]
     runs-on: ${{ matrix.os }}
     steps:
     - uses: cachix/install-nix-action@v20

--- a/.github/workflows/nix.yml
+++ b/.github/workflows/nix.yml
@@ -10,7 +10,7 @@ jobs:
         os: [ ubuntu-latest, macos-latest ]
     runs-on: ${{ matrix.os }}
     steps:
-    - uses: cachix/install-nix-action@v20
+    - uses: cachix/install-nix-action@v22
       with:
         github_access_token: ${{ secrets.GITHUB_TOKEN }}
     - uses: cachix/cachix-action@v12

--- a/default.nix
+++ b/default.nix
@@ -79,6 +79,6 @@ stdenv.mkDerivation {
     maintainers = with lib.maintainers; [ inclyc ];
     platforms = lib.platforms.unix;
     # Tracking issue: https://github.com/nix-community/nixd/issues/107
-    broken = stdenv.isDarwin;
+    # broken = stdenv.isDarwin;
   };
 }

--- a/default.nix
+++ b/default.nix
@@ -78,7 +78,5 @@ stdenv.mkDerivation {
     license = lib.licenses.lgpl3Plus;
     maintainers = with lib.maintainers; [ inclyc ];
     platforms = lib.platforms.unix;
-    # Tracking issue: https://github.com/nix-community/nixd/issues/107
-    # broken = stdenv.isDarwin;
   };
 }

--- a/flake.nix
+++ b/flake.nix
@@ -12,12 +12,7 @@
     perSystem = { config, self', inputs', pkgs, system, ... }:
       with pkgs;
       let
-        nixd = callPackage ./default.nix {
-          stdenv =
-            if stdenv.isDarwin
-            then llvmPackages_16.libcxxStdenv
-            else stdenv;
-        };
+        nixd = callPackage ./default.nix { };
       in
       {
         packages.default = nixd;
@@ -29,14 +24,10 @@
         devShells.default = nixd.overrideAttrs (old: {
           nativeBuildInputs = old.nativeBuildInputs ++ (with pkgs; [
             clang-tools
-            gdb
             nixpkgs-fmt
-            llvmPackages_16.clang
           ]);
           shellHook = ''
             export PATH="${pkgs.clang-tools}/bin:$PATH"
-            export NIX_SRC=${pkgs.nixUnstable.src}
-            export NIX_DEBUG_INFO_DIRS=${pkgs.nixUnstable.debug}/lib/debug
           '';
         });
 

--- a/flake.nix
+++ b/flake.nix
@@ -13,6 +13,9 @@
       with pkgs;
       let
         nixd = callPackage ./default.nix { };
+        nixdLLVM = nixd.override {
+          stdenv = if stdenv.isDarwin then stdenv else llvmPackages.stdenv;
+        };
       in
       {
         packages.default = nixd;
@@ -20,6 +23,8 @@
           inherit (config.packages) nixd;
         };
         packages.nixd = nixd;
+
+        devShells.llvm = nixdLLVM;
 
         devShells.default = nixd.overrideAttrs (old: {
           nativeBuildInputs = old.nativeBuildInputs ++ (with pkgs; [

--- a/lib/lspserver/src/Logger.cpp
+++ b/lib/lspserver/src/Logger.cpp
@@ -101,7 +101,9 @@ StreamLogger::StreamLogger(llvm::raw_ostream &Logs, Logger::Level MinLevel)
   pthread_mutexattr_t Attr;
   pthread_mutexattr_init(&Attr);
   pthread_mutexattr_setpshared(&Attr, PTHREAD_PROCESS_SHARED);
+#ifndef __APPLE__
   pthread_mutexattr_setrobust(&Attr, PTHREAD_MUTEX_ROBUST);
+#endif
   pthread_mutex_init(ShmLock, &Attr);
   pthread_mutexattr_destroy(&Attr);
 }

--- a/lib/nixd/src/ServerController.cpp
+++ b/lib/nixd/src/ServerController.cpp
@@ -30,7 +30,6 @@
 #include <memory>
 #include <mutex>
 #include <optional>
-#include <ranges>
 #include <sstream>
 #include <stdexcept>
 #include <string>
@@ -368,7 +367,11 @@ void Server::onDecalration(const lspserver::TextDocumentPositionParams &Params,
     std::string Punc = "\r\n\t ;";
 
     auto IsPunc = [&Punc](char C) {
-      return std::ranges::any_of(Punc, [C](char Ck) { return Ck == C; });
+      for (const auto Char : Punc) {
+        if (Char == C)
+          return true;
+      }
+      return false;
     };
 
     for (; From >= 0 && !IsPunc(Code[From]);)

